src/js_ast.zig

```diff
@@ -36,189 +36,153 @@ const MimeType = bun.http.MimeType;
-
-            if (overflow.used == 0) {
-                if (overflow.allocated == 0 or overflow.ptrs[0].used == 0) {
-                    return &.{};
-                }
-            }
+        const PreAlloc = struct {
+            metadata: Store,
+            first_block: Block,
+        };
 
-            var to_move = overflow.ptrs[0..overflow.allocated][overflow.used..];
+        pub fn init(backing_allocator: Allocator) *Store {
+            const prealloc = backing_allocator.create(PreAlloc) catch bun.outOfMemory();
 
-            // This returns the list of maxed out blocks
-            var used_list = overflow.slice();
+            prealloc.first_block.bytes_used = 0;
+            prealloc.first_block.prev = null;
+            prealloc.first_block.next = null;
 
-            // The last block may be partially used.
-            if (overflow.allocated > overflow.used and to_move.len > 0 and to_move.ptr[0].used > 0) {
-                to_move = to_move[1..];
-                used_list.len += 1;
-            }
+            prealloc.metadata = .{
+                .backing_allocator = backing_allocator,
+                .head = &prealloc.first_block,
+                .current = &prealloc.first_block,
+            };
 
-            const used = overflow.allocator.dupe(*Block, used_list) catch unreachable;
+            return &prealloc.metadata;
+        }
 
-            for (to_move, overflow.ptrs[0..to_move.len]) |b, *out| {
-                b.* = Block{
-                    .items = undefined,
-                    .used = 0,
-                };
-                out.* = b;
+        pub fn deinit(store: *Store) void {
+            var it = store.head.next; // do not free `store.head`
+            while (it) |next| {
+                if (Environment.isDebug)
+                    @memset(next.buffer, undefined);
+                it = next.next;
+                store.backing_allocator.destroy(next);
             }
 
-            overflow.allocated = @as(Overflow.UsedSize, @truncate(to_move.len));
-            overflow.used = 0;
-
-            return used;
+            const prealloc: PreAlloc = @fieldParentPtr("metadata", store);
+            bun.assert(&prealloc.first_block == store.head);
         }
 
-        /// Reset all AST nodes, allowing the memory to be reused for the next parse.
-        /// Only call this when we're done with ALL AST nodes, or you risk
-        /// undefined memory bugs.
-        ///
-        /// Nested parsing should either use the same store, or call
-        /// Store.reclaim.
-        pub fn reset() void {
-            const blocks = _self.?.overflow.slice();
-            for (blocks) |b| {
-                if (comptime Environment.isDebug) {
-                    // ensure we crash if we use a freed value
-                    const bytes = std.mem.asBytes(&b.items);
-                    @memset(bytes, undefined);
+        pub fn reset(store: *Store) void {
+            store.current = store.head;
```

Comment:
Didn't use a linked list for the previous implemnetation to make iterating through the complete list when recycling fast

Maybe it could have a list of 1000 node pointers batched into one allocation?